os/kernel/semaphore/sem_destroy.c

```diff
@@ -85,6 +88,27 @@
 /****************************************************************************
  * Private Functions
  ****************************************************************************/
+#if defined(CONFIG_BINMGR_RECOVERY) && defined(__KERNEL__)
+extern sq_queue_t g_sem_list;
+
+static void sem_unregister(FAR sem_t *sem)
+{
+	sem_t *sem_ptr;
+	irqstate_t flags;
+
+	flags = irqsave();
+	sem_ptr = (sem_t *)sq_peek(&g_sem_list);
+	while (sem_ptr) {
+		if (sem_ptr == sem) {
+			/* Remove semaphore from a list of kernel semaphore */
+			sq_rem((FAR sq_entry_t *)sem, &g_sem_list);
```

Comment:
sq_rem searches the correct sem.
So we don't need to run the while loop for searching.